A. Build a baseline model (5 marks) 

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.
1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [1]:
#Downloading the dataset 
#importing the packages
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')  #Reading the dataset using pandas
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.isnull().sum() #checking for any missing values

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [5]:
concrete_data_columns = concrete_data.columns

X = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
y = concrete_data['Strength'] # Strength column

Splitting the data into training and testing sets using train_test_split function

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

Saving the number of predictors

In [7]:
n_cols = X.shape[1] #n_cols is the number of predictors now

Starting to Build the model

In [8]:
#importing Keras and other packages 
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [9]:
from keras.models import Sequential
from keras.layers import Dense

Defining the Regression Model

In [10]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape = (n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Model training

In [11]:
# build the model
model = regression_model()

In [12]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=0)

In [13]:
y_pred = model.predict(X_test)

from sklearn.metrics import mean_squared_error
MSE = mean_squared_error(y_pred, y_test)
print(MSE)

120.8444801016533


Now to calculate it 50 times more we add a loop

In [14]:
#making an array to store MSE values
MSE = []

In [15]:
for k in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    MSE.append(mean_squared_error(y_pred, y_test))

In [16]:
from statistics import mean, stdev

meanMSE = mean(MSE)
sdMSE = stdev(MSE)

print("The mean of MSE is %.2f" % meanMSE)
print("The standard deviation of MSE is %.2f" % sdMSE)

The mean of MSE is 322.12
The standard deviation of MSE is 339.95


B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [17]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [18]:
#normalizing the data
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Now repeating A

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3)

In [20]:
n_cols = predictors_norm.shape[1] #n_cols is the number of predictors now

In [21]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape = (n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [22]:
# build the model
model = regression_model()

In [23]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=0)

In [24]:
y_pred = model.predict(X_test)

from sklearn.metrics import mean_squared_error
MSE = mean_squared_error(y_pred, y_test)
print(MSE)

290.7185151248783


In [25]:
#making an array to store MSE values
MSE = []

In [26]:
for k in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30)
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    MSE.append(mean_squared_error(y_pred, y_test))

In [27]:
from statistics import mean, stdev

meanMSE = mean(MSE)
sdMSE = stdev(MSE)

print("The mean of MSE is %.2f" % meanMSE)
print("The standard deviation of MSE is %.2f" % sdMSE)

The mean of MSE is 370.44
The standard deviation of MSE is 123.27


In part A the values for mean and stand deviation of MSE were :
The mean of MSE is 322.12
The standard deviation of MSE is 339.95

C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [30]:
#repeating Part B but with 100 epochs
for k in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30)
    model = regression_model()
    model.fit(X_train, y_train, epochs=100, verbose=0)
    y_pred = model.predict(X_test)
    MSE.append(mean_squared_error(y_pred, y_test))

In [31]:
from statistics import mean, stdev

meanMSE = mean(MSE)
sdMSE = stdev(MSE)

print("The mean of MSE is %.2f" % meanMSE)
print("The standard deviation of MSE is %.2f" % sdMSE)

The mean of MSE is 301.29
The standard deviation of MSE is 131.71


In part B the values for mean and standard deviation of MSE were :
The mean of MSE is 370.44
The standard deviation of MSE is 123.27.

D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [34]:
#redefining the regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape = (n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [35]:
# build the model
model = regression_model()

In [36]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=0)

In [37]:
y_pred = model.predict(X_test)

from sklearn.metrics import mean_squared_error
MSE = mean_squared_error(y_pred, y_test)
print(MSE)

155.07838836699642


In [38]:
#making an array to store MSE values
MSE = []

In [39]:
for k in range(0, 50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30)
    model = regression_model()
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    MSE.append(mean_squared_error(y_pred, y_test))

In [40]:
from statistics import mean, stdev

meanMSE = mean(MSE)
sdMSE = stdev(MSE)

print("The mean of MSE is %.2f" % meanMSE)
print("The standard deviation of MSE is %.2f" % sdMSE)

The mean of MSE is 132.74
The standard deviation of MSE is 16.81


In part B the values for mean and standard deviation of MSE were : The mean of MSE is 370.44 The standard deviation of MSE is 123.27.